### PR TITLE
Update source list when back and forward buttons are clicked

### PR DESCRIPTION
**Pull request by arthur801031**
Associated Issue: #635

### Summary of Changes
* Reload debuggee when a back/forward button is clicked. 

### Test Plan
1. Open debugger in Chrome
2. Open debuggee in Firefox
3. Go to https://devtools-html.github.io/debugger-examples/examples/todomvc/ in the debugee
4. Go to https://www.cnet.com/
5. Observe Sources
6. Click back button 
7. Observe Sources
8. Click forward button
9. Observe Sources

### CODE REVIEW

File: src/actions/navigation.js [hunk 1]
```diff
@@ -23,15 +25,13 @@ function willNavigate() {
...
-    // TODO: This seems to be buggy on the debugger server side. When
-    // the page is loaded from bfcache, we still get sources from the
-    // *previous* page as well. For now, emulate the current debugger
-    // behavior by not showing sources loaded by bfcache.
-    // return dispatch(sources.loadSources());
+  return ({ dispatch, getState }: ThunkArgs) => {
+    setTimeout(() => {
+      if (getSources(getState()).size == 0) {
+        const threadClient = firefox.getThreadClient();
+        threadClient.getSources();
```
Comment (jasonLaster) [inline src/actions/navigation.js]: Looking good.

1. does this fix the problem, my fear is that getSources will make one call for the sources, but it won't trigger `nweSource` by itself
2. i think we can refactor this code a bit to make it more consistent with `util/client`

https://gist.github.com/jasonLaster/de3bd2ecbacaf7518c42336e199d5137
https://gist.github.com/dde3f2cd62733746e562f9d6acd48e76
 
Goal:
1. add a new firefox client command for fetching sources
2. add a `timeout` helper so we can do it with async/await
3. call `newSources` in this action which i think we need to do

File: src/utils/utils.js [hunk 2]
```diff
@@ -184,7 +184,7 @@ function updateObj<T: Object>(obj: T, fields: $Shape<T>) : T {
  * @static
  */
 function throttle(func: any, ms: number) {
-  let timeout, _this;
+  let timeout, _this; // eslint-disable-line no-shadow
```
Comment (arthur801031) [inline src/utils/utils.js]: I got this error, which I thought it's fine to disable it.

screen shot 2017-02-01 at 12 49 48 am

**Comment by arthur801031** [comment]
Hi @jasonLaster 

I've made those changes. The PR for the refactoring is [here](https://github.com/devtools-html/devtools-core/pull/149). Thank you! :)

**Comment by jasonLaster** [comment]
I'll pull in the client work soon

File: src/utils/utils.js (hunk 2 above)
Comment (jasonLaster) [inline src/utils/utils.js]: hmm, maybe we want to change the function name to `waitForMs`

File: src/actions/navigation.js [hunk 3]
```diff
@@ -23,15 +26,14 @@ function willNavigate() {
...
-  return ({ dispatch }) => {
-    // We need to load all the sources again because they might have
-    // come from bfcache, so we won't get a `newSource` notification.
-    //
-    // TODO: This seems to be buggy on the debugger server side. When
-    // the page is loaded from bfcache, we still get sources from the
-    // *previous* page as well. For now, emulate the current debugger
-    // behavior by not showing sources loaded by bfcache.
-    // return dispatch(sources.loadSources());
+  return ({ dispatch, client, getState }) => {
```
Comment (jasonLaster) [inline src/actions/navigation.js]: We have a slightly nicer style here:
https://github.com/arthur801031/debugger.html/blob/update_source/src/actions/pause.js#L45-L48

**Pull request merged**